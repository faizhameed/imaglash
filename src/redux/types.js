const actionTypes = {
  UNSPLASH_INITIALISED: "UNSPLASH_INITIALISED",
  ON_PAGE_LOAD: "ON_PAGE_LOAD",
  SEARCH_QUERY_CHANGED: "SEARCH_QUERY_CHANGED",
  LOAD_MORE_DATA: "lOAD_MORE_DATA",
  RANDOM_IMAGE_TRIGGER: "RANDOM_IMAGE_TRIGGER",
  USER_CLICKS_IMAGE: "USER_CLICKS_IMAGE",
};

export default actionTypes;
