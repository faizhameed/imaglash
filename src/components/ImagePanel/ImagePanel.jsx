import React, { useEffect } from "react";
import Button from "../Button/Button";

import {
  updatePageNumber,
  setCollection,
  userSelectImage,
} from "../../redux/actions";
import { connect } from "react-redux";

import "./ImagePanel.scss";
import { Link } from "react-router-dom";

const ImagePanel = ({
  updatePageNumber,
  collection,
  setCollection,
  page,
  currentQuery,
  userSelectImage,
}) => {
  useEffect(() => {
    setCollection();
  }, [setCollection]);

  return (
    <React.Fragment>
      <div className="image-container">
        {collection.length !== 0
          ? collection.map((item) => (
              <Link key={item.id} to="/image-view">
                <div
                  className="image-element"
                  onClick={() => userSelectImage(item)}
                >
                  <img src={item.urls.small} alt={item.alt_description} />
                  <div className="user-detail-sm">
                    <img
                      src={item.user.profile_image.medium}
                      alt={item.user.name}
                    />
                    <p>
                      Image By <span>{item.user.name}</span>
                    </p>
                  </div>
                </div>
              </Link>
            ))
          : null}
      </div>
      <Button onClick={() => updatePageNumber(page, currentQuery)}>
        Load More
      </Button>
    </React.Fragment>
  );
};

const mapStateToProps = ({
  unsplashReducer: { collection, page, currentQuery },
}) => ({
  collection,
  page,
  currentQuery,
});

const mapDispatchToProps = (dispatch) => ({
  updatePageNumber: (page, query) => dispatch(updatePageNumber(page, query)),
  setCollection: () => dispatch(setCollection()),
  userSelectImage: (data) => dispatch(userSelectImage(data)),
});
export default connect(mapStateToProps, mapDispatchToProps)(ImagePanel);
