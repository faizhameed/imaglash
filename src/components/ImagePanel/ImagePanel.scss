.image-container {
  display: flex;
  flex-wrap: wrap;
  .image-element {
    margin: 10px;
    border-radius: 4px;
  }
}
