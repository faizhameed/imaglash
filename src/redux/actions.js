import actionTypes from "./types";

import Unsplash, { toJson } from "unsplash-js";

const unsplash = new Unsplash({
  accessKey: "C6bzEnmvODyk9wJ3ig0V4E7p9pwGVkZNejOftxLLdVI",
});

export const changeSearchQuery = (query) => {
  return function (dispatch) {
    if (unsplash) {
      unsplash.search
        .photos(query, 1, 9)
        .then(toJson)
        .then((json) => {
          dispatch({
            type: actionTypes.SEARCH_QUERY_CHANGED,
            payload: json,
            query: query,
          });
        });
    }
  };
};

export const updatePageNumber = (page, query) => {
  return function (dispatch) {
    if (unsplash) {
      unsplash.search
        .photos(query, page + 1, 9)
        .then(toJson)
        .then((json) => {
          dispatch({
            type: actionTypes.LOAD_MORE_DATA,
            payload: json,
            query,
          });
        });
    }
  };
};

export const setCollection = () => {
  return function (dispatch) {
    if (unsplash) {
      unsplash.search
        .photos("book", 1, 9)
        .then(toJson)
        .then((json) => {
          dispatch({
            type: actionTypes.ON_PAGE_LOAD,
            payload: json,
          });
        });
    }
  };
};
