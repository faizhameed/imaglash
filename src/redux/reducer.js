import actionTypes from "./types";
import { combineReducers } from "redux";

const UNSPLASH_STATE = {
  unsplash: null,
  tags: [
    "Dog",
    "Cat",
    "Space",
    "Nature",
    "Business",
    "Office",
    "Coffee",
    "World",
    "Wildlife",
    "Beach",
    "Digital",
    "Meeting",
    "Cars",
    "Games",
    "Holiday",
  ],
  currentQuery: "Book",
  page: 1,
  collection: [],
  totalPages: 0,
  randomImageUrl: "",
  userImageSelection: null,
};

export function unsplashReducer(state = UNSPLASH_STATE, action) {
  switch (action.type) {
    case actionTypes.UNSPLASH_INITIALISED:
      return {
        ...state,
        unsplash: action.payload,
      };
    case actionTypes.ON_PAGE_LOAD:
      return {
        ...state,
        collection: [...action.payload.results],
        totalPages: action.payload.total_pages,
        page: 1,
      };

    case actionTypes.SEARCH_QUERY_CHANGED:
      return {
        ...state,

        collection: [...action.payload.results],
        page: 1,
        totalPages: action.payload.total_pages,
        currentQuery: action.query,
      };

    case actionTypes.LOAD_MORE_DATA:
      if (state.page < state.totalPages)
        return {
          ...state,
          page: state.page + 1,
          collection: [...state.collection, ...action.payload.results],
        };
      break;
    case actionTypes.RANDOM_IMAGE_TRIGGER:
      return {
        ...state,
        randomImageUrl: action.payload,
      };

    case actionTypes.USER_CLICKS_IMAGE:
      return {
        ...state,
        userImageSelection: action.payload,
      };
    default:
      return state;
  }
}

const rootReducer = combineReducers({
  unsplashReducer,
});

export default rootReducer;
