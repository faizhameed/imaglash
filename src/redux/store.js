import { createStore } from "redux";
import rootReducer from "./reducer";
import { applyMiddleware } from "redux";
import { createLogger } from "redux-logger";
import thunkMiddleware from "redux-thunk";

const logger = createLogger();
/**
 * we are creating global store and using the store in our root file to have
 * all components to access it
 *
 */

const middlewares = [thunkMiddleware];
/**
 * we need not run logger in production mode since it can show the state to users.
 * To improve security of out application we should only run logger in development mode
 */
if (process.env.NODE_ENV === "development") {
  middlewares.push(logger);
}
export const store = createStore(rootReducer, applyMiddleware(...middlewares));
