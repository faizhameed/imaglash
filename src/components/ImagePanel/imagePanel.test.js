import React from "react";
import { shallow } from "enzyme";
import ImagePanel from "./ImagePanel";
import { Provider } from "react-redux";
import { store } from "../../redux/store";

describe("Image Panel Component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Provider store={store}>
        <ImagePanel />
      </Provider>
    );
  });
  it("expect to render Image panel", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
