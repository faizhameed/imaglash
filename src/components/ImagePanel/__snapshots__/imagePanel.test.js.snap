// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Panel Component expect to render Image panel 1`] = `ShallowWrapper {}`;
